feat(picker): make the telescope harpoon-picker behave more like a builtin telescope file-picker

kimabrandt-flx opened the pull request:
This commit fixes the missing `entry.index` - in the marks-extension for telescope - which caused the index to always show `nil` - or rather `n…` - due to a fixed width of 2, in the dialogue.

This commit also added a dynamic width and left-padding for the index-column, so the index is being right-aligned.

kimabrandt-flx committed:
fix(marks): fix missing entry-index

This commit fixes the missing `entry.index` - in the marks-extension for
telescope - which caused the index to always show `nil` - or rather
`n…` - due to a fixed width of 2, in the dialogue.

This commit also added a dynamic width and left-padding for the
index-column, so the index is being right-aligned.

kimabrandt-flx committed:
feat(picker): behave more like a telescope builtin file-picker

These changes allow the telescope harpoon-picker to behave more like a
builtin telescope file-picker. Both global and individual telescope-
customizations will have an effect.

kimabrandt-flx committed:
feat(picker): add icons to the harpoon file-picker

kimabrandt-flx committed:
fix(picker): give the row and column their own place, in the display-config

kimabrandt-flx committed:
feat(picker): enhance the usability of the marks-picker

Improve the `move`- and `delete`-operations, by making them feel
smoother. Also make the movements behave more like "builtin" telescope-
movements.

kimabrandt-flx commented:
These changes add some enhancements for the telescope `Harpoon Marks`-picker.

kimabrandt-flx committed:
fix(marks): icrement the `harpoon_item.context.col` by 1, for display

kimabrandt-flx committed:
feat(picker): sort by index

This change allows the telescope-picker to sort by index. If one of the
search-terms equals an index, it'll get a better score. Partial matches
of an index, will get a slightly better score. Matches with the main
entry - i.e. the filename - will always have the best score; showing
them before an index-match.

gvolpe commented:
This is great, thank you for putting up this work!

gvolpe commented on lua/telescope/_extensions/marks.lua:
```diff
@@ -1,135 +1,284 @@
-local move_mark_up = function(prompt_bufnr)
-    local selection = action_state.get_selected_entry()
-    local length = harpoon:list():length()
+local delete_mark_selections_prompt = function(prompt_bufnr)
```
nitpick: it would be great if this could be configurable (i.e. being able to disable this prompt and delete the mark straight away)

kimabrandt-flx replied:
I could also imagine to override the mappings!? Something like this:

```lua
require("telescope").extensions.harpoon.marks(themes.get_dropdown({
    previewer = false,
    layout_config = { width = 0.6 },
    path_display = { truncate = 10 },
    attach_mappings = function(_, map)
        map("i", "<c-d>", hm_actions.delete_mark_selections)
        map("n", "<c-d>", hm_actions.delete_mark_selections)
        return true
    end,
}))
```

The different actions (e.g. `delete_mark_selections`, `delete_mark_selections_prompt`) would've to be exported first!?

For now, you could just override the mappings in the code and use `delete_mark_selections` instead of `delete_mark_selections_prompt`!

gvolpe replied:
That would be even better! :)

Indeed, the actions need to be exported first. I'm using your fork now, thanks for doing this again 🙏🏽

kimabrandt-flx replied:
You can now do:

```lua
local themes = require("telescope.themes")
local hm_actions = require("telescope._extensions.harpoon_marks.actions")
require("telescope").extensions.harpoon.marks(themes.get_dropdown({
    previewer = false,
    layout_config = { width = 0.6 },
    path_display = { truncate = 10 },
    attach_mappings = function(_, map)
        map("i", "<c-d>", hm_actions.delete_mark_selections)
        map("n", "<c-d>", hm_actions.delete_mark_selections)
        return true
    end,
}))
```

ThePrimeagen commented:
If I can get a couple of approvals, I will merge this.

Not know the telescope stuff, so I need someone who can take the lead on reviewing these changes

kimabrandt-flx committed:
refactor: remove unused require

kimabrandt-flx committed:
refactor(picker): modularize the marks-picker

This can be used to extend and override the key-mappings inside
the `attach_mappings`-function, for example:

```lua
local themes = require("telescope.themes")
local hm_actions = require("telescope._extensions.harpoon_marks.actions")
require("telescope").extensions.harpoon.marks(themes.get_dropdown({
    previewer = false,
    layout_config = { width = 0.6 },
    path_display = { truncate = 10 },
    attach_mappings = function(_, map)
        map("i", "<c-d>", hm_actions.delete_mark_selections)
        map("n", "<c-d>", hm_actions.delete_mark_selections)
        return true
    end,
}))
```

kimabrandt-flx committed:
Merge branch 'harpoon2' into fix_marks_index

kimabrandt-flx committed:
fix: rename `harpoon:list():removeAt` into `harpoon:list():remove_at`

kimabrandt-flx committed:
Merge branch 'harpoon2' into fix_marks_index

gvolpe reviewed (approved):
@ThePrimeagen I'm not a Telescope expert, but all I can say is this works perfectly! Thank you again @kimabrandt-flx 🙏🏽

11Firefox11 commented:
Tried it out with [this](https://github.com/ThePrimeagen/harpoon/pull/512#discussion_r1552372809) code. I don't know telescope, I just tried it out and <a href="https://github.com/ThePrimeagen/harpoon/pull/512#issuecomment-2206108308"><s>it works well</s></a>. 

One thing I noticed though is that if you delete an item then the Harpoon indexes will change too, but in telescope the original numbers persist. For example in Telescope: 1, 2, 3, 4, 5 --delete 2nd--> 1, 3, 4, 5 (should be 1, 2, 3, 4 - which it will be if I close and reopen the whole thing).
Sadly I don't know how to update the currently displayed items in telescope so I can't implement this.

11Firefox11 commented:
I used it for few days and it doesn't work well. The ordering really breaks when deleting items and then adding new ones. There will be empty rows in the original Harpoon window.

kimabrandt-flx commented:
@11Firefox11
The current Telescope marks-extension, has the same issue: https://github.com/ThePrimeagen/harpoon/issues/588
Maybe this could be fixed inside the [remove_at](https://github.com/ThePrimeagen/harpoon/blob/0378a6c428a0bed6a2781d459d7943843f374bce/lua/harpoon/list.lua#L205)-function!? Calling it, seems to leave a blank value (`nil`) inside the harpoon-list!? I'm not sure if the Telescope-picker should show the blank value, or if the blank value should be removed, inside the harpoon-list?!

I think - when restarting - you could be seeing something related to these issues!? https://github.com/ThePrimeagen/harpoon/issues/476, https://github.com/ThePrimeagen/harpoon/pull/573
